feat(core): implement context propagation and optimize service architecture

s0up4200 opened the pull request:
## Context Propagation
- Implemented consistent context flow across all service layers
- Enhanced resource cleanup and timeout management
- Updated database and cache operations to respect parent contexts
- Improved error propagation and handling

## Service Architecture Optimization
- Refactored service layer for better reliability and performance
- Implemented singleflight pattern to prevent redundant operations
- Enhanced service configuration loading and initialization
- Improved health monitoring stability

## SSE Implementation
- Optimized Server-Sent Events (SSE) across services
- Enhanced update frequency management for different services:
  - Plex: Improved session updates with SSE support
  - Overseerr: Added SSE support with optimized cache durations
  - Autobrr: Added top 5 releases to card with SSE updates
  - *arr services: Improved version update handling

## Performance Enhancements
- Optimized cache layer operations
- Improved service health check efficiency
- Enhanced connection handling and stability
- Reduced redundant data fetching

## UI/UX Improvements
- Autobrr: Added tooltips for download and info links
- Autobrr: Improved media type detection and icon handling
- Plex: Enhanced progress bar behavior
- Updated CSP for better development experience (hopefully this stops the crashes in dev and make the app more usable on http etc)

# Technical Details
- All service layers now properly propagate contexts
- Database operations integrated with parent contexts
- Cache operations optimized with proper context handling
- API layer updated with improved error handling
- Health monitoring system refactored for better stability
- SSE implementation optimized across all services

# Testing
- Updated cache and database integration tests
- Improved service health monitoring tests
- Added context propagation tests
- Verified SSE performance and stability

s0up4200 committed:
perf: optimize service configuration loading and initialization

s0up4200 committed:
fix(sse): optimize health updates to prevent double reloads

- add debouncing and batching for SSE health updates
- memoize status counts to prevent unnecessary re-renders
- improve initial load and configuration change handling
- add proper cleanup for update timeouts

This fixes the UI reload issue by optimizing how SSE updates are processed
and reducing unnecessary re-renders while maintaining real-time updates.

s0up4200 committed:
perf(plex): improve session update frequency with SSE support

s0up4200 committed:
feat(overseerr): add SSE support and optimize cache durations

s0up4200 committed:
added the missing services and increased timeouts

s0up4200 committed:
fix: sse

s0up4200 committed:
feat: make overseer look better

s0up4200 committed:
Merge branch 'develop' into fix/optimize-sse-updates

s0up4200 committed:
Merge branch 'develop' into fix/optimize-sse-updates

s0up4200 committed:
fix(arrs): show version update as warning msg

s0up4200 committed:
feat: add autobrr to sse and add top 5 releases to card

KyleSanderson commented on internal/api/handlers/autobrr.go:
```diff
@@ -183,6 +317,37 @@ func (h *AutobrrHandler) fetchAndCacheStats(instanceId, cacheKey string) (autobr
 	return stats, nil
 }
 
+func (h *AutobrrHandler) fetchAndCacheReleases(instanceId, cacheKey string) (autobrr.ReleasesResponse, error) {
```
https://pkg.go.dev/golang.org/x/sync/singleflight

s0up4200 committed:
feat(services): implement context propagation across services

Adds context propagation throughout the service layer to improve request lifecycle management and cancellation support. This change:

- Updates service interfaces to include context parameter
- Implements context-aware methods in all services
- Maintains compatibility with arr.HealthChecker interface
- Ensures proper context propagation in handlers and background tasks
- This improves request handling by allowing proper timeout and cancellation management, while maintaining backward compatibility with existing arr services.

s0up4200 committed:
refactor(database): update methods to accept context parameter

Commit description:
Update database layer to properly handle context propagation throughout all operations. This change is part of Phase 2 of the context propagation refactor.

Key changes:

Add context parameter to all database methods
Replace internal context.Background() calls with passed contexts
Update SQL operations to use context-aware methods (QueryContext, ExecContext)
Add context cancellation tests
Update documentation to reflect completed Phase 2

s0up4200 committed:
refactor(cache): update cache layer to use passed contexts

This commit updates the cache layer to properly handle context propagation:

- Add context parameter to cache constructors:
  - NewCache(ctx context.Context, addr string)
  - NewMemoryStore(ctx context.Context, dataDir string)
  - InitCache(ctx context.Context, cfg Config)
- Remove internal context.Background() usage
- Update all callers to pass context appropriately
- Add context cancellation handling
- Update tests to verify context handling
- Maintain backward compatibility

s0up4200 committed:
refactor(api): update API layer to use proper context handling

Key Changes:
- Update handlers to use Request.Context() with timeouts:
  - Health handler: Add context timeout for health checks
  - Auth handler: Add timeouts for oauth2 and HTTP operations
  - Improve error handling for context cancellation

- Update middleware for better context handling:
  - Add context timeouts for auth operations
  - Add custom context keys for session data
  - Improve context propagation through middleware chain
  - Add proper error handling for timeouts

- Key improvements:
  - Replace gin.Context with Request.Context()
  - Add appropriate timeouts for different operations
  - Add cleanup on context cancellation
  - Improve error responses for timeouts

s0up4200 committed:
fix: improve health monitoring stability and SSE connection handling

- Fix SSE connection handling in development mode by properly configuring Vite proxy
- Reduce concurrent health checks from 10 to 5 to prevent overwhelming services
- Implement batched health checks (3 services per batch) with proper timeouts
- Add automatic cleanup of stale connections and force reconnect for long-running ones
- Improve error handling and resource management in health check system
- Add proper documentation of health monitoring architecture and patterns

This addresses crashes and connection issues in development mode by preventing resource exhaustion and handling slow/unresponsive services more gracefully.
The changes ensure better stability through proper connection management and cleanup mechanisms.

s0up4200 committed:
fix(tests): update cache and database integration tests

s0up4200 committed:
feat(handlers): implement singleflight pattern across remaining handlers

Implements singleflight pattern to prevent duplicate requests and cache stampede:
- Added singleflight to Autobrr, Radarr, Sonarr, Omegabrr, Overseerr, Plex, Prowlarr and Tailscale handlers
- Protected main operations and cache refresh operations
- Maintained consistent error handling and SSE broadcasting
- Preserved existing caching patterns and response formats

This change reduces load on backend services, improves response times for
concurrent requests, and ensures memory efficiency with automatic cleanup.

s0up4200 committed:
fix: hide healthy prowlarr status msg

s0up4200 committed:
refactor(autobrr): move structs into types package

s0up4200 committed:
fix(autobrr): hide irc status if healthy

s0up4200 committed:
feat(autobrr): add tooltips for download and info links

s0up4200 committed:
fix(web): type-checks

s0up4200 committed:
fix(plex): prevent progress bar flash with zero offset

s0up4200 committed:
fix(middleware): update CSP to allow local development connections

s0up4200 committed:
feat(autobrr): improve media type detection and icon handling

- Create centralized media type detection system
- Add comprehensive category mapping for various tracker types
- Implement proper TypeScript types for HeroIcon components
- Move media type logic to dedicated utility file
- Simplify AutobrrStats component by removing duplicate logic
